primitives/relay/src/lib.rs

```diff
@@ -8,238 +8,127 @@ use core::fmt::Debug;
 
-	fn hash(&self) -> Self::Hash;
+	fn store_relay_parcel(relay_parcel: Self::RelayParcel) -> DispatchResult;
 }
 
-// A regulator to adjust relay args for a specific chain
-// Implement this in runtime's impls
+/// A regulator to adjust relay args for a specific chain
+/// Implement this in runtime's `impls.rs`
 pub trait AdjustableRelayerGame {
 	type Moment;
 	type Balance;
-	type TcBlockNumber;
+	type RelayBlockId;
 
-	fn challenge_time(round: Round) -> Self::Moment;
+	/// The maximum number of active games
+	///
+	/// This might relate to the validators count
+	fn max_active_games() -> u8;
 
-	fn round_of_samples_count(samples_count: u64) -> Round;
+	fn propose_time(round: u32) -> Self::Moment;
 
-	fn samples_count_of_round(round: Round) -> u64;
+	fn complete_proofs_time(round: u32) -> Self::Moment;
 
-	fn update_samples(samples: &mut Vec<Vec<Self::TcBlockNumber>>);
+	/// Update the samples
+	///
+	/// Push the new samples to the `samples`, the index of `samples` aka round index
+	/// And return the new samples
+	fn update_samples(samples: &mut Vec<Vec<Self::RelayBlockId>>);
 
-	fn estimate_bond(round: Round, proposals_count: u64) -> Self::Balance;
+	/// Give an estimate bond value for a specify round
+	///
+	/// Usally the bond value go expensive wihle the round and the proposals count increase
+	fn estimate_bond(round: u32, proposals_count: u8) -> Self::Balance;
 }
 
 pub trait RelayerGameProtocol {
 	type Relayer;
-	type Balance;
-	type HeaderThingWithProof;
-	type HeaderThing: HeaderThing;
-
-	fn proposals_of_game(
-		game_id: <Self::HeaderThing as HeaderThing>::Number,
-	) -> Vec<
-		RelayProposal<
-			Self::Relayer,
-			Self::Balance,
-			Self::HeaderThing,
-			<Self::HeaderThing as HeaderThing>::Hash,
-		>,
-	>;
-
-	fn submit_proposal(
+	type GameId: Clone + PartialOrd;
+	type RelayParcel: Clone + Debug + PartialEq + PartialOrd + FullCodec;
+	type Proofs;
+
+	/// Game's entry point, call only at the first round
+	///
+	/// Propose a new proposal or against a existed proposal
+	fn propose(
 		relayer: Self::Relayer,
-		proposal: Vec<Self::HeaderThingWithProof>,
+		game_id: Self::GameId,
+		relay_parcel: Self::RelayParcel,
+		proofs: Option<Self::Proofs>,
+	) -> DispatchResult;
+
+	/// Verify a specify proposal
+	///
+	/// Proofs is a `Vec` because the sampling function might give more than 1 sample points,
+	/// so need to verify each sample point with its proofs
+	fn complete_proofs(
+		proposal_id: RelayProposalId<Self::GameId>,
+		proofs: Vec<Self::Proofs>,
 	) -> DispatchResult;
 
-	fn approve_pending_header(
-		pending: <Self::HeaderThing as HeaderThing>::Number,
+	/// Once there're different opinions in a game,
+	/// chain will ask relayer to submit more samples
+	/// to help the chain make a on chain arbitrate finally
+	fn extend_proposal(
+		relayer: Self::Relayer,
+		samples: Vec<Self::RelayParcel>,
+		extended_relay_proposal_id: RelayProposalId<Self::GameId>,
+		proofses: Option<Vec<Self::Proofs>>,
 	) -> DispatchResult;
```

Comment:
Still considering using `RelayParcelWithProof` here

```rust
struct RelayParcelWithProof {
  parcel: RelayParcel,
  proofs: Option<Proofs>
}

fn extend_proposal(
  relayer: Self::Relayer,
  samples: Vec<Self::RelayParcelWithProof>,
  extended_relay_proposal_id: RelayProposalId<Self::Proofs>
)
```

Fewer arguments might be better for the design of the API.

Reply:
I prefer without a struct, but the 'proofses' look weird, maybe 'proofs_list' or just 'proofs', a proof is a proof of a sample.

Reply:
That's what exactly the API's job.

If there's `option_proofs` then I just need one `if let Some()`.

But if there's a `Vec<RelayParcelWithProof>` then I've to `for relay_pracel_with_proof in samples { if let Some() }` for every sigle option filed.